File: app/src/main/java/com/lambdaschool/swapi/SwApiDao.java
package com.lambdaschool.swapi;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;

public class SwApiDao {
    public interface ObjectCallback<T> {
        void returnPlanets(T object);
    }

    public static void getAllPlanets(final ObjectCallback<ArrayList<Planet>> objectCallback) {
        final ArrayList<Planet> planets = new ArrayList<>();
        final NetworkAdapter.NetworkCallback callback = new NetworkAdapter.NetworkCallback() {
            @Override
            public void returnResult(Boolean success, String page) {
                // process page of data
                String nextUrl = null;
                try {
                    nextUrl = new JSONObject(page).getString("next");
                } catch (JSONException e) {
                    e.printStackTrace();
                    nextUrl = null;
                }
                // yay recursion!
                if (nextUrl != null) {
                    NetworkAdapter.httpGetRequest(nextUrl, this);
                }

                try {
                    JSONObject pageJson     = new JSONObject(page);
                    JSONArray  resultsArray = pageJson.getJSONArray("results");
                    for (int i = 0; i < resultsArray.length(); ++i) {
                        try {
                            planets.add(new Planet(resultsArray.getJSONObject(i)));
                        } catch (JSONException e) {
                            e.printStackTrace();
                        }
                    }
                } catch (JSONException e) {
                    e.printStackTrace();
                }

                if (nextUrl == null) {
                    /*synchronized (planets) {
                        planets.notify();
                    }*/
                    objectCallback.returnPlanets(planets);
                }
            }
        };
        NetworkAdapter.httpGetRequest("https://swapi.co/api/planets", callback);
    }

    public static void getAllTransports(final ObjectCallback<Transport> objectCallback) {
        final ArrayList<Transport> transports = new ArrayList<>();
        NetworkAdapter.httpGetRequest("https://swapi.co/api/vehicles", new NetworkAdapter.NetworkCallback() {
            @Override
            public void returnResult(Boolean success, String page) {
                // process page of data
                String nextUrl = "https://swapi.co/api/vehicles";
                try {
                    JSONObject pageJson     = new JSONObject(page);
                    JSONArray  resultsArray = pageJson.getJSONArray("results");
                    for (int i = 0; i < resultsArray.length(); ++i) {
                        try {
                            transports.add(new Vehicle(resultsArray.getJSONObject(i)));
                            objectCallback.returnPlanets(transports.get(transports.size() - 1));
                        } catch (JSONException e) {
                            e.printStackTrace();
                        }
                    }
                } catch (JSONException e) {
                    e.printStackTrace();
                }

                try {
                    nextUrl = new JSONObject(page).getString("next");
                    NetworkAdapter.httpGetRequest(nextUrl, this);
                } catch (JSONException e) {
                    e.printStackTrace();
                    nextUrl = null;
                }
            }
        });

        NetworkAdapter.httpGetRequest("https://swapi.co/api/starships", new NetworkAdapter.NetworkCallback() {
            @Override
            public void returnResult(Boolean success, String page) {
                // process page of data
                String nextUrl;
                try {
                    JSONObject pageJson     = new JSONObject(page);
                    JSONArray  resultsArray = pageJson.getJSONArray("results");
                    for (int i = 0; i < resultsArray.length(); ++i) {
                        try {
                            transports.add(new Starship(resultsArray.getJSONObject(i)));
                            objectCallback.returnPlanets(transports.get(transports.size() - 1));
                        } catch (JSONException e) {
                            e.printStackTrace();
                        }
                    }
                } catch (JSONException e) {
                    e.printStackTrace();
                }

                try {
                    nextUrl = new JSONObject(page).getString("next");
                } catch (JSONException e) {
                    e.printStackTrace();
                    nextUrl = null;
                }
            }
        });
    }
}

File: app/src/main/java/com/lambdaschool/swapi/SwApiDao.kt
package com.lambdaschool.swapi

import android.util.Log

import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

import java.util.ArrayList
import java.util.concurrent.Semaphore
import java.util.concurrent.atomic.AtomicBoolean

object SwApiDao {
    interface ObjectCallback<T> {
        fun returnPlanets(`object`: T)
    }

    fun getAllPlanets(canceled: AtomicBoolean, objectCallback: ObjectCallback<ArrayList<Planet>>) {
        val planets = ArrayList<Planet>()
        val lock = Semaphore(1)
        val callback = object : NetworkAdapter.NetworkCallback {
            override fun returnResult(success: Boolean?, page: String) {
                // process page of data
                if (canceled.get()) {
                    Log.i("GetRequestCanceled", page)
                    return
                }

                var nextUrl: String? = null
                try {
                    nextUrl = JSONObject(page).getString("next")
                } catch (e: JSONException) {
                    e.printStackTrace()
                    nextUrl = null
                }

                // yay recursion!
                if (nextUrl != null) {
                    NetworkAdapter.httpGetRequest(nextUrl, canceled, this)
                }

                try {
                    val pageJson = JSONObject(page)
                    val resultsArray = pageJson.getJSONArray("results")

                    if (canceled.get()) {
                        Log.i("GetRequestCanceled", page)
                        return
                    }

                    for (i in 0 until resultsArray.length()) {
                        try {
                            lock.acquire()
                            planets.add(Planet(resultsArray.getJSONObject(i)))
                            lock.release()
                        } catch (e: JSONException) {
                            e.printStackTrace()
                        } catch (e: InterruptedException) {
                            e.printStackTrace()
                        }

                    }
                } catch (e: JSONException) {
                    e.printStackTrace()
                }

                if (nextUrl == null) {
                    /*synchronized (planets) {
                        planets.notify();
                    }*/
                    objectCallback.returnPlanets(planets)
                }
            }
        }
        if (canceled.get()) {
            Log.i("GetRequestCanceled", "First")
            return
        }
        NetworkAdapter.httpGetRequest("https://swapi.co/api/planets", canceled, callback)
    }

    fun getAllTransports(objectCallback: ObjectCallback<Transport>) {
        val transports = ArrayList<Transport>()
        val lock = Semaphore(1)
        NetworkAdapter.httpGetRequest("https://swapi.co/api/vehicles", object : NetworkAdapter.NetworkCallback {
            override fun returnResult(success: Boolean?, page: String) {
                // process page of data
                var nextUrl: String? = "https://swapi.co/api/vehicles"
                try {
                    val pageJson = JSONObject(page)
                    val resultsArray = pageJson.getJSONArray("results")
                    for (i in 0 until resultsArray.length()) {
                        try {
                            lock.acquire()
                            transports.add(Vehicle(resultsArray.getJSONObject(i)))
                            objectCallback.returnPlanets(transports[transports.size - 1])
                            lock.release()
                        } catch (e: JSONException) {
                            e.printStackTrace()
                        } catch (e: InterruptedException) {
                            e.printStackTrace()
                        }

                    }
                } catch (e: JSONException) {
                    e.printStackTrace()
                }

                try {
                    nextUrl = JSONObject(page).getString("next")
                    NetworkAdapter.httpGetRequest(nextUrl, this)
                } catch (e: JSONException) {
                    e.printStackTrace()
                    nextUrl = null
                }

            }
        })

        NetworkAdapter.httpGetRequest("https://swapi.co/api/starships") { success, page ->
            // process page of data
            var nextUrl: String?
            try {
                val pageJson = JSONObject(page)
                val resultsArray = pageJson.getJSONArray("results")
                for (i in 0 until resultsArray.length()) {
                    try {
                        lock.acquire()
                        transports.add(Starship(resultsArray.getJSONObject(i)))
                        objectCallback.returnPlanets(transports[transports.size - 1])
                        lock.release()
                    } catch (e: JSONException) {
                        e.printStackTrace()
                    } catch (e: InterruptedException) {
                        e.printStackTrace()
                    }

                }
            } catch (e: JSONException) {
                e.printStackTrace()
            }

            try {
                nextUrl = JSONObject(page).getString("next")
            } catch (e: JSONException) {
                e.printStackTrace()
                nextUrl = null
            }
        }
    }
}
